ts/private/ts_project.bzl

```diff
@@ -182,83 +182,111 @@ See https://github.com/aspect-build/rules_ts/issues/361 for more details.
-    use_tsc_noemit = ctx.attr.no_emit or (ctx.attr.transpile == 0 and not ctx.attr.declaration)
+
+    # What tsc will be emitting
+    use_tsc_for_js = len(js_outs) > 0
+    use_tsc_for_dts = len(typings_outs) > 0
+
+    # Use a separate non-emitting action for type-checking when:
+    #  - a isolated typechecking action was explicitly requested
+    #  or
+    #  - not invoking tsc for output files at all
+    use_isolated_typecheck = ctx.attr.isolated_typecheck or not (use_tsc_for_js or use_tsc_for_dts)
+
+    # We don't produce any DefaultInfo outputs in this case, because we avoid running the tsc action
+    # unless the output_types are requested.
+    default_outputs = []
 
     # Default outputs (DefaultInfo files) is what you see on the command-line for a built
     # library, and determines what files are used by a simple non-provider-aware downstream
     # library. Only the JavaScript outputs are intended for use in non-TS-aware dependents.
-    if ctx.attr.transpile != 0:
+    if use_tsc_for_js:
         # Special case case where there are no source outputs and we don't have a custom
         # transpiler so we add output_types to the default outputs
         default_outputs = output_sources if len(output_sources) else output_types
-    else:
-        # We must avoid tsc writing any JS files in this case, as tsc was only run for typings, and some other
-        # action will try to write the JS files. We must avoid collisions where two actions write the same file.
-        if not use_tsc_noemit:
-            arguments.add("--emitDeclarationOnly")
-
-        # We don't produce any DefaultInfo outputs in this case, because we avoid running the tsc action
-        # unless the output_types are requested.
-        default_outputs = []
 
     srcs_tsconfig_deps = ctx.attr.srcs + [ctx.attr.tsconfig] + ctx.attr.deps
 
-    stdout_file = ""
+    inputs = copy_files_to_bin_actions(ctx, inputs)
 
-    if use_tsc_noemit:
+    transitive_inputs.append(_gather_types_from_js_infos(srcs_tsconfig_deps))
+    transitive_inputs_depset = depset(
+        inputs,
+        transitive = transitive_inputs,
+    )
+
+    # tsc action for type-checking
+    if use_isolated_typecheck:
         # The type-checking action still need to produce some output, so we output the stdout
         # to a .typecheck file that ends up in the typecheck output group.
         typecheck_output = ctx.actions.declare_file(ctx.attr.name + ".typecheck")
         typecheck_outs.append(typecheck_output)
 
-        outputs.append(typecheck_output)
-        stdout_file = typecheck_output.path
-
+        typecheck_arguments = ctx.actions.args()
         if supports_workers:
-            arguments.add("--bazelValidationFile", typecheck_output.short_path)
+            typecheck_arguments.add("--bazelValidationFile", typecheck_output.short_path)
 
-        arguments.add("--noEmit")
+        typecheck_arguments.add("--noEmit")
```

Comment:
When `support_workers` is `True` on line 108 we call `arguments.use_param_file`, in which case I think we'd want `"--noEmit"` to be added to that params file, rather than a command-line argument.

Reply:
Ohhh, you mean instead of `typecheck_arguments` it must be `arguments`?

Reply:
Yeah, though it seems tricky since `arguments` is also used down below where we wouldn't want it to have `--noEmit`. As far as I can tell there's not an easy way to create a copy of an `Args` object or remove an element. Something like this might be able to help with that: https://github.com/bazelbuild/bazel/issues/6230, but it looks stalled.

Reply:
Pretty dirty, but something like this seems to work
```diff
diff --git a/ts/private/ts_project.bzl b/ts/private/ts_project.bzl
index df88bc6..71c9ff3 100644
--- a/ts/private/ts_project.bzl
+++ b/ts/private/ts_project.bzl
@@ -74,7 +74,7 @@ def _ts_project_impl(ctx):
 
     typecheck_outs = []
 
-    arguments = ctx.actions.args()
+    argument_mods = []
     execution_requirements = {}
     executable = ctx.executable.tsc
 
@@ -105,34 +105,35 @@ See https://github.com/aspect-build/rules_ts/issues/361 for more details.
 
     if supports_workers:
         # Set to use a multiline param-file for worker mode
-        arguments.use_param_file("@%s", use_always = True)
-        arguments.set_param_file_format("multiline")
+        argument_mods.append(lambda arguments: arguments.use_param_file("@%s", use_always = True))
+        argument_mods.append(lambda arguments: arguments.set_param_file_format("multiline"))
         execution_requirements["supports-workers"] = "1"
         execution_requirements["worker-key-mnemonic"] = "TsProject"
         executable = ctx.executable.tsc_worker
 
     # Add all arguments from options first to allow users override them via args.
-    arguments.add_all(options.args)
+    argument_mods.append(lambda arguments: arguments.add_all(options.args))
 
     # Add user specified arguments *before* rule supplied arguments
-    arguments.add_all(ctx.attr.args)
+    argument_mods.append(lambda arguments: arguments.add_all(ctx.attr.args))
 
     outdir = _lib.join(
         ctx.label.workspace_root,
         _lib.join(ctx.label.package, ctx.attr.out_dir) if ctx.attr.out_dir else ctx.label.package,
     )
     tsconfig_path = to_output_relative_path(tsconfig)
-    arguments.add_all([
+    argument_mods.append(lambda arguments: arguments.add_all([
         "--project",
         tsconfig_path,
         "--outDir",
         outdir,
         "--rootDir",
         _lib.calculate_root_dir(ctx),
-    ])
+    ]))
+
     if len(typings_outs) > 0:
         declaration_dir = _lib.join(ctx.label.workspace_root, ctx.label.package, typings_out_dir)
-        arguments.add("--declarationDir", declaration_dir)
+        argument_mods.append(lambda arguments:  arguments.add("--declarationDir", declaration_dir))
 
     inputs = srcs_inputs + tsconfig_inputs
 
@@ -156,7 +157,7 @@ See https://github.com/aspect-build/rules_ts/issues/361 for more details.
 
     outputs = js_outs + map_outs + typings_outs + typing_maps_outs
     if ctx.outputs.buildinfo_out:
-        arguments.add("--tsBuildInfoFile", to_output_relative_path(ctx.outputs.buildinfo_out))
+        argument_mods.append(lambda arguments: arguments.add("--tsBuildInfoFile", to_output_relative_path(ctx.outputs.buildinfo_out)))
         outputs.append(ctx.outputs.buildinfo_out)
 
     output_sources = js_outs + map_outs + assets_outs
@@ -222,16 +223,19 @@ See https://github.com/aspect-build/rules_ts/issues/361 for more details.
         typecheck_output = ctx.actions.declare_file(ctx.attr.name + ".typecheck")
         typecheck_outs.append(typecheck_output)
 
-        typecheck_arguments = ctx.actions.args()
-        if supports_workers:
-            typecheck_arguments.add("--bazelValidationFile", typecheck_output.short_path)
+        arguments = ctx.actions.args()
+        for mod in argument_mods:
+            mod(arguments)
 
-        typecheck_arguments.add("--noEmit")
+        arguments.add("--noEmit")
+
+        if supports_workers:
+            arguments.add("--bazelValidationFile", typecheck_output.short_path)
 
         ctx.actions.run(
             executable = executable,
             inputs = transitive_inputs_depset,
-            arguments = [arguments, typecheck_arguments],
+            arguments = [arguments],
             outputs = [typecheck_output],
             mnemonic = "TsProjectCheck",
             execution_requirements = execution_requirements,
@@ -249,6 +253,10 @@ See https://github.com/aspect-build/rules_ts/issues/361 for more details.
         typecheck_outs.extend(output_types)
 
     if use_tsc_for_js or use_tsc_for_dts:
+        arguments = ctx.actions.args()
+        for mod in argument_mods:
+            mod(arguments)
+
         tsc_emit_arguments = ctx.actions.args()
 
         # Type-checking is done async as a separate action and can be skipped.
```

Reply:
^ 
https://github.com/aspect-build/rules_ts/blob/4a4bf962b758cc5c754834c1b069c234178131df/ts/private/ts_project_worker.js#L997-L999

Notably, `--bazelValidationFile` is also expected to be in the params file